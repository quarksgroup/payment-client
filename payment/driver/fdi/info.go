package fdi

import (
	"context"
	"fmt"
	"time"

	"github.com/iradukunda1/payment-staging/payment/mtn"
)

type infoService struct {
	client *wrapper
}

func (s *infoService) Info(ctx context.Context, ref string) (*mtn.Info, *mtn.Response, error) {
	endpoint := fmt.Sprintf("momo/trx/%s/info", ref)
	out := new(infoResponse)
	res, err := s.client.do(ctx, "GET", endpoint, nil, out)
	return convertInfo(out), res, err
}

type infoResponse struct {
	Status string `json:"status"`
	Data   struct {
		ID         string    `json:"id"`
		Ref        string    `json:"trxRef"`
		Type       string    `json:"trxType"`
		ChannelID  string    `json:"channelId"`
		ChannelRef string    `json:"channelRef"`
		MSISDN     string    `json:"msisdn"`
		Amount     float64   `json:"amount"`
		Fees       float64   `json:"trxFees"`
		Currency   string    `json:"currency"`
		TrxStatus  string    `json:"trxStatus"`
		CreatedAt  time.Time `json:"created_at"`
		Callback   string    `json:"callback"`
	} `json:"data"`
}

func convertInfo(info *infoResponse) *mtn.Info {
	return &mtn.Info{
		ID:        info.Data.Ref,
		Amount:    info.Data.Amount,
		Cost:      info.Data.Fees,
		Status:    info.Data.TrxStatus,
		Type:      info.Data.Type,
		CreatedAt: info.Data.CreatedAt,
	}
}
