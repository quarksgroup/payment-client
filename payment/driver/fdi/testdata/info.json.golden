{
  "id": "19567751-F53B-4B15-A4B4",
  "amount": 100,
  "cost": 2,
  "status": "successful",
  "type": "cashin",
  "createdAt": "2019-04-04 17:15:59"
}
